src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,587 @@
+/// written with a StartTime or BaseTime, that metadata will appear in header comments, and that
+/// will be represented by the iterator providing the corresponding variants here. The presence
+/// of those timestamps will affect how you should interpret the timestamps for individual
+/// intervals. See the module-level documentation.
+#[allow(variant_size_differences)]
+pub enum LogEntry<'a> {
+    /// Logs may include a StartTime. If present, it represents seconds since the epoch.
+    StartTime(f64),
+    /// Logs may include a BaseTime. If present, it represents seconds since the epoch.
+    BaseTime(f64),
+    /// An individual interval histogram.
+    Interval(IntervalLogHistogram<'a>),
+}
+
+/// Errors that occur when parsing an interval log.
+#[derive(Debug, PartialEq)]
+pub enum LogIteratorError {
+    /// Parsing failed
+    ParseError {
+        /// Offset in the input where the failed parse started
+        offset: usize,
+    },
+}
+
+/// Parse interval logs.
+///
+/// This iterator exposes each item (excluding comments and other information-free lines). See
+/// `LogEntry`.
+///
+/// Because histogram deserialization is deferred, parsing logs is fast. See the `interval_log`
+/// benchmark if you wish to see how it does on your hardware. As a baseline, parsing a log of 1000
+/// random histograms of 10,000 values each takes 8ms total on an E5-1650v3.
+///
+/// Deferring deserialization is handy because it allows you to cheaply navigate the log to find
+/// the records you care about (e.g. ones in a certain time range, or with a certain tag) without
+/// doing all the allocation, etc, of deserialization.
+///
+/// If you're looking for a direct port of the Java impl's `HistogramLogReader`, this isn't one: it
+/// won't deserialize for you, and it pushes the burden of figuring out what to do with StartTime,
+/// BaseTime, etc to you, and there aren't built in functions to filter by timestamp. On the other
+/// hand, because it doesn't do those things, it is much more flexible: you can easily build any
+/// sort of filtering you want, not just timestamp ranges, because you have cheap access to all the
+/// metadata before incurring the cost of deserialization. If you're not using any timestamp
+/// headers, or at least using them in straightforward ways, it is easy to accumulate the
+/// timestamp state you need. Since all the parsing is taken care of already, writing your own
+/// `HistogramLogReader` equivalent that fits the way your logs are assembled is just a couple of
+/// lines. (And if you're doing complex stuff, we probably wouldn't have built something that fits
+/// your quirky logs anyway!)
+///
+/// This parses from a slice representing the complete file because it made implementation easier
+/// (and also supports mmap'd files for maximum parsing speed). If parsing from a `Read` is
+/// important for your use case, open an issue about it.
```

Comment:
This is an interesting point. I suspect parsing from a `Read` will be a very useful addition, and one that will be tricky to combine with the way the code is currently written, but I agree that we can deal with that on-demand.

Reply:
I don't think it will be too bad; we'll just have to read into a buffer, parse what we can, and grow the buffer if we get `Incomplete`. There's some streaming support in `nom`, but I didn't investigate too closely. More than I felt like tackling on top of learning `nom`. :)